src/scope_traits.rs

```diff
@@ -0,0 +1,168 @@
+use crate::scope::Entered;
+use crate::CallbackScope;
+use crate::Context;
+use crate::EscapableHandleScope;
+use crate::FunctionCallbackInfo;
+use crate::HandleScope;
+use crate::Isolate;
+use crate::Local;
+use crate::Locker;
+use crate::Message;
+use crate::Object;
+use crate::PropertyCallbackInfo;
+use crate::ReturnValue;
+
+pub(crate) mod internal {
+  use super::*;
+
+  extern "C" {
+    fn v8__Context__GetIsolate(self_: &Context) -> *mut Isolate;
+    fn v8__EscapableHandleScope__GetIsolate(
+      self_: &EscapableHandleScope,
+    ) -> *mut Isolate;
+    fn v8__FunctionCallbackInfo__GetIsolate(
+      self_: &FunctionCallbackInfo,
+    ) -> *mut Isolate;
+    fn v8__HandleScope__GetIsolate(self_: &HandleScope) -> *mut Isolate;
+    fn v8__Message__GetIsolate(self_: &Message) -> *mut Isolate;
+    fn v8__Object__GetIsolate(self_: &Object) -> *mut Isolate;
+    fn v8__PropertyCallbackInfo__GetIsolate(
+      self_: &PropertyCallbackInfo,
+    ) -> *mut Isolate;
+    fn v8__ReturnValue__GetIsolate(self_: &ReturnValue) -> *mut Isolate;
+  }
+
+  /// Internal trait for retrieving a raw Isolate pointer from various V8
+  /// API objects.
+  pub trait GetRawIsolate {
+    fn get_raw_isolate(&self) -> *mut Isolate;
+  }
+
+  impl<'s, T> GetRawIsolate for Local<'s, T>
+  where
+    Local<'s, Object>: From<Local<'s, T>>,
+  {
+    fn get_raw_isolate(&self) -> *mut Isolate {
+      let local = Local::<'s, Object>::from(*self);
+      (&*local).get_raw_isolate()
+    }
+  }
+
+  impl<'s> GetRawIsolate for Local<'s, Context> {
+    fn get_raw_isolate(&self) -> *mut Isolate {
+      (&**self).get_raw_isolate()
+    }
+  }
+
+  impl<'s> GetRawIsolate for Local<'s, Message> {
+    fn get_raw_isolate(&self) -> *mut Isolate {
+      (&**self).get_raw_isolate()
```

Comment:
I don't understand why there are two stars here...

Reply:
The first * calls `<Local<T> as Deref>::deref` to `&Message`.
The second * calls `<Message as Deref>::deref` to get to its base class (`Object`).